File: lib/services/payload_generator.dart
class PayloadGeneratorOptions {
  final String url;
  final String method; // GET, POST, CONNECT, etc.
  final String injectionMethod; // Normal, Front Inject, Back Inject, Dual Real, Split
  final bool keepAlive;
  final bool userAgent;
  final bool upgradeWebsocket;
  final bool customHeader;
  final bool referer;
  final bool forwardedHost;
  final bool backQuery;
  final bool frontQuery;

  PayloadGeneratorOptions({
    this.url = 'id1.jagoanip.my.id',
    this.method = 'GET',
    this.injectionMethod = 'Normal',
    this.keepAlive = false,
    this.userAgent = false,
    this.upgradeWebsocket = true,
    this.customHeader = false,
    this.referer = false,
    this.forwardedHost = false,
    this.backQuery = false,
    this.frontQuery = false,
  });
}

class PayloadGenerator {
  /// Parses tags inside the payload string and replaces them with appropriate values.
  static String parsePayload(
    String payload, {
    required String host,
    required int port,
    String statusLine = 'HTTP/1.1 200 Connection established',
  }) {
    String parsed = payload;

    parsed = parsed.replaceAll('[crlf]', '\r\n');
    parsed = parsed.replaceAll('[cr]', '\r');
    parsed = parsed.replaceAll('[lf]', '\n');
    parsed = parsed.replaceAll('[protocol]', 'HTTP/1.1');
    parsed = parsed.replaceAll('[host]', host);
    parsed = parsed.replaceAll('[port]', port.toString());
    parsed = parsed.replaceAll(
      '[host_port]',
      port == 80 || port == 443 ? host : '$host:$port',
    );
    parsed = parsed.replaceAll(
      '[ua]',
      'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
    );
    parsed = parsed.replaceAll(
      '[raw]',
      'CONNECT [host_port] [protocol][crlf]Host: [host_port][crlf]',
    );
    parsed = parsed.replaceAll(
      '[real_raw]',
      'CONNECT [host_port] [protocol][crlf]',
    );
    parsed = parsed.replaceAll('[status]', statusLine);

    return parsed;
  }

  /// Generates a standard HTTP payload string from user options.
  static String generatePayload(PayloadGeneratorOptions options) {
    String host = options.url.trim();
    if (host.isEmpty) host = 'id1.jagoanip.my.id';

    StringBuffer sb = StringBuffer();

    if (options.injectionMethod == 'Front Inject') {
      sb.write('${options.method} http://$host/ HTTP/1.1[crlf]');
      sb.write('Host: $host[crlf]');
    } else if (options.injectionMethod == 'Back Inject') {
      sb.write('CONNECT [host_port] HTTP/1.1[crlf]');
      sb.write('Host: [host_port][crlf]');
      sb.write('${options.method} http://$host/ HTTP/1.1[crlf]');
    } else {
      // Normal
      sb.write('${options.method} / HTTP/1.1[crlf]');
      sb.write('Host: $host[crlf]');
    }

    if (options.upgradeWebsocket) {
      sb.write('Upgrade: websocket[crlf]');
    }

    if (options.keepAlive) {
      sb.write('Connection: Keep-Alive[crlf]');
    }

    if (options.userAgent) {
      sb.write('User-Agent: [ua][crlf]');
    }

    if (options.referer) {
      sb.write('Referer: http://$host/[crlf]');
    }

    if (options.forwardedHost) {
      sb.write('X-Forwarded-Host: $host[crlf]');
    }

    sb.write('[crlf]');
    return sb.toString();
  }
}

File: app/src/main/kotlin/com/sslh/sshl/app/service/PayloadGenerator.kt
package com.sslh.sshl.app.service

data class PayloadGeneratorOptions(
    val url: String = "id1.jagoanip.my.id",
    val method: String = "GET",
    val injectionMethod: String = "Normal",
    val keepAlive: Boolean = false,
    val userAgent: Boolean = false,
    val upgradeWebsocket: Boolean = true,
    val customHeader: Boolean = false,
    val referer: Boolean = false,
    val forwardedHost: Boolean = false,
    val backQuery: Boolean = false,
    val frontQuery: Boolean = false,
    val onlineHost: Boolean = false,
    val reverseProxy: Boolean = false,
    val dualConnect: Boolean = false
)

object PayloadGenerator {

    fun parsePayload(
        payload: String,
        host: String,
        port: Int,
        statusLine: String = "HTTP/1.1 200 Connection established"
    ): String {
        val hostPort = if (port == 80 || port == 443) host else "$host:$port"
        val userAgentStr = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36"

        var parsed = payload
        parsed = parsed.replace("[crlf]", "\r\n")
        parsed = parsed.replace("[cr]", "\r")
        parsed = parsed.replace("[lf]", "\n")
        parsed = parsed.replace("[lfcr]", "\n\r")
        parsed = parsed.replace("[crlf*2]", "\r\n\r\n")
        parsed = parsed.replace("[protocol]", "HTTP/1.1")
        parsed = parsed.replace("[host]", host)
        parsed = parsed.replace("[port]", port.toString())
        parsed = parsed.replace("[host_port]", hostPort)
        parsed = parsed.replace("[ua]", userAgentStr)
        parsed = parsed.replace("[method]", "CONNECT")
        parsed = parsed.replace("[auth]", "")
        parsed = parsed.replace("[netData]", "CONNECT $hostPort HTTP/1.1\r\nHost: $hostPort\r\n\r\n")
        parsed = parsed.replace("[raw]", "CONNECT $hostPort HTTP/1.1\r\nHost: $hostPort\r\n")
        parsed = parsed.replace("[real_raw]", "CONNECT $hostPort HTTP/1.1\r\n")
        parsed = parsed.replace("[status]", statusLine)
        parsed = parsed.replace("[instant]", "")

        return parsed
    }

    fun generatePayload(options: PayloadGeneratorOptions): String {
        var host = options.url.trim()
        if (host.isEmpty()) host = "id1.jagoanip.my.id"

        val queryParam = if (options.frontQuery) "http://$host/" else if (options.backQuery) "?$host" else ""
        val targetPath = if (options.frontQuery) "http://$host/" else "/$queryParam"

        val sb = StringBuilder()

        if (options.dualConnect) {
            sb.append("CONNECT [host_port] HTTP/1.1[crlf]")
        }

        when (options.injectionMethod) {
            "Front Inject" -> {
                sb.append("${options.method} http://$host/ HTTP/1.1[crlf]")
                sb.append("Host: $host[crlf]")
            }
            "Back Inject" -> {
                sb.append("CONNECT [host_port] HTTP/1.1[crlf]")
                sb.append("Host: [host_port][crlf]")
                sb.append("${options.method} http://$host/ HTTP/1.1[crlf]")
            }
            else -> {
                // Normal
                sb.append("${options.method} $targetPath HTTP/1.1[crlf]")
                sb.append("Host: $host[crlf]")
            }
        }

        if (options.onlineHost) {
            sb.append("X-Online-Host: $host[crlf]")
        }

        if (options.reverseProxy) {
            sb.append("X-Forward-Host: $host[crlf]")
        }

        if (options.upgradeWebsocket) {
            sb.append("Upgrade: websocket[crlf]")
        }

        if (options.keepAlive) {
            sb.append("Connection: Keep-Alive[crlf]")
        }

        if (options.userAgent) {
            sb.append("User-Agent: [ua][crlf]")
        }

        if (options.referer) {
            sb.append("Referer: http://$host/[crlf]")
        }

        if (options.forwardedHost) {
            sb.append("X-Forwarded-Host: $host[crlf]")
        }

        sb.append("[crlf]")
        return sb.toString()
    }
}
